UI/window-basic-preview.cpp

```diff
@@ -1091,6 +1136,170 @@ static void DrawCircleAtPos(float x, float y, matrix4 &matrix,
 	gs_matrix_pop();
 }
 
+static void DrawLine(vec3 &startPos, vec3 &endPos)
+{
+	struct gs_vb_data *vbData = gs_vbdata_create();
+	vbData->num = 2;
+	vbData->points = (vec3*)bzalloc(sizeof(struct vec3) * vbData->num);
+	vbData->normals = (vec3*)bzalloc(sizeof(struct vec3) * vbData->num);
+	vbData->colors = (uint32_t*)bzalloc(sizeof(uint32_t) * vbData->num);
+
+	vbData->points[0] = startPos;
+	vbData->points[1] = endPos;
+
+	vbData->num_tex = 1;
+	vbData->tvarray = (gs_tvertarray*)bzalloc(sizeof(struct gs_tvertarray));
+	vbData->tvarray[0].width = 2;
+	vbData->tvarray[0].array = bzalloc(sizeof(struct vec2) * vbData->num);
+
+	gs_vertbuffer_t *vb = gs_vertexbuffer_create(vbData, GS_DYNAMIC);
+	gs_load_vertexbuffer(vb);
+
+	gs_draw(GS_LINES, 0, 0);
+	gs_vertexbuffer_destroy(vb);
+}
+
+static void DrawLabel(vec3 &pos, obs_source_t *source, vec3 &viewport) {
+	if (!source)
+		return;
+
+	vec3_mul(&pos, &pos, &viewport);
+
+	gs_matrix_push();
+	gs_matrix_identity();
+	gs_matrix_translate(&pos);
+	obs_source_video_render(source);
+	gs_matrix_pop();
+}
+
+void OBSBasicPreview::DrawSingleSpacingHelper(vec3 &start, vec3 &end,
+	vec3 &viewport)
+{
+	const float labelMargin = 0.005f;
+	const float virtualLabelSizeFactor = 1.05f;
+
+	obs_video_info ovi;
+	obs_get_video_info(&ovi);
+
+	float length, lengthPx;
+	length = sqrt(pow(end.x - start.x, 2) + pow(end.y - start.y, 2));
+	if ((end.x - start.x) >= (end.y - start.y)) {
+		// Horizontal line
+		lengthPx = length * ovi.base_width;
+	} else {
+		// Vertical line
+		lengthPx = length * ovi.base_height;
+	}
+
+	std::string lengthStr;
+	lengthStr = std::to_string((int)lengthPx) + " px";
+
+	if (!sizeLabels[currentSizeLabel]) {
+		sizeLabels[currentSizeLabel] = CreateLabel("Spacing Helper");
+	}
+	obs_source_t* sizeLabel = sizeLabels[currentSizeLabel];
```

Comment:
I tried using a single text source, by changing text and calling obs_source_video_render right after, doing this consecutively for each label, but this didn't work. Apparently the actual render to texture is not done when calling obs_source_video_render.